File: 10_valid_number/main.c
#include <stdio.h>
#include "stdbool.h"

bool isDigit(char s) {
    return s >= '0' && s <= '9';
}

bool isNumber(char *s) {
    //states
    int NUMBER = 1, DOT = 2, E = 3, SIGN = 4, SPACE = 5;

    int current_state = 0;

    bool got_dot = false, got_e = false, got_numbers = false;

    for (int i = 0; s[i] != '\0'; i++) {
        // if the current state is 0 then we just started or only found white spaces so far
        if (s[i] == ' ') {
            if (current_state == 1) {
                current_state = SPACE;
                continue;
            } else
                continue;
        }

        // initial state
        if (current_state == 0) {
            // check if we got a number
            if (isDigit(s[i])) {
                current_state = NUMBER;
                got_numbers = true;
                continue;
            } else if (s[i] == '.') {
                current_state = DOT;
                got_dot = true;
                continue;
            } else if (s[i] == '+' || s[i] == '-') {
                current_state = SIGN;
                continue;
            } else
                return false;
        }
            // if the state is NUMBER
        else if (current_state == NUMBER) {
            if (isDigit(s[i])) {
                got_numbers = true;
                continue;
            } else if (s[i] == '.' && !got_dot) {
                current_state = DOT;
                got_e = true;
                continue;
            } else if (s[i] == 'e' && !got_e) {
                current_state = E;
                got_e = true;
                continue;
            } else
                return false;
        }
            // if the state is DOT
        else if (current_state == DOT) {
            if (isDigit(s[i])) {
                got_numbers = true;
                current_state = NUMBER;
                continue;
            } else if (s[i] == 'e' && got_numbers) {
                got_e = true;
                current_state = E;
                continue;
            } else
                return false;
        }
            // if the state is E
        else if (current_state == E) {
            // can not have any more dots so, lets lie and say that we found a dot
            got_dot = true;
            if (isDigit(s[i])) {
                got_numbers = true;
                current_state = NUMBER;
                continue;
            } else if (s[i] == '+' || s[i] == '-') {
                current_state = SIGN;
                continue;
            } else
                return false;
        }
            // if the state is SIGN
        else if (current_state == SIGN) {
            if (isDigit(s[i])) {
                got_numbers = true;
                current_state = NUMBER;
                continue;
            } else if (s[i] == '.' && !got_dot) {
                current_state = DOT;
                got_dot = true;
                continue;
            } else
                return false;
        }
            // if the state is SPACE
        else if (current_state == SPACE) {
            if (s[i] != ' ')
                return false;
        }
    }

    if (current_state != SPACE && current_state != NUMBER && current_state != DOT || !got_numbers)
        return false;

    return true;
}

int main() {
    printf("%d", isNumber(". 9"));
    return 0;
}

File: 10_valid_number/main.py
class Solution:
    def isNumber(self, s: str) -> bool:
        # this problem was solved by using a finite state machine

        # states
        NUMBER = 1
        DOT = 2
        E = 3
        SIGN = 4

        current_state = 0

        got_dot = False
        got_e = False
        got_numbers = False

        s = s.strip()

        for c in s:
            # first iteration
            if current_state == 0:
                if c.isdigit():
                    current_state = NUMBER
                    got_numbers = True
                    continue
                elif c == '.':
                    current_state = DOT
                    got_dot = True
                    continue
                elif c == '+' or c == '-':
                    current_state = SIGN
                    continue
                else:
                    return False

            if current_state == NUMBER:
                if c.isdigit():
                    got_numbers = True
                    continue
                elif c == '.' and got_dot is False:
                    current_state = DOT
                    got_dot = True
                    continue
                elif c == 'e' and got_e is False:
                    current_state = E
                    got_e = True
                    continue
                else:
                    return False
            elif current_state == DOT:
                if c.isdigit():
                    got_numbers = True
                    current_state = NUMBER
                    continue
                elif c == 'e' and got_numbers is True:
                    got_e = True
                    current_state = E
                    continue
                else:
                    return False
            elif current_state == E:
                # can not have a decimal value after the e
                got_dot = True
                if c.isdigit():
                    got_numbers = True
                    current_state = NUMBER
                    continue
                elif c == '+' or c == '-':
                    current_state = SIGN
                    continue
                else:
                    return False
            elif current_state == SIGN:
                if c.isdigit():
                    got_numbers = True
                    current_state = NUMBER
                    continue
                elif c == '.' and got_dot is False:
                    current_state = DOT
                    got_dot = True
                    continue
                else:
                    return False

        # if the last state is not the state NUMBER or the DOT then return false
        if current_state != NUMBER and current_state != DOT or got_numbers is False:
            return False

        return True


num = '.1e21 '
print(Solution().isNumber(num))  # True

num = '0.1'
print(Solution().isNumber(num))  # True

num = '1abc1'
print(Solution().isNumber(num))  # False

num = '2e10'
print(Solution().isNumber(num))  # True

num = '6e-1'
print(Solution().isNumber(num))  # True

num = '99e2.5'
print(Solution().isNumber(num))  # False
